### PR TITLE
[#165180769] handling login spid webview errors

### PR DESCRIPTION
**Pull request by Undermaken**
Sometime could happen the login request on SPID provider will fail: no internet connection, http error, timeout etc.
So we need to give a feedback to the user and a way to retry the failed request

### CODE REVIEW

**Comment by cloudify** [comment]
This pull request **introduces 1 alert** when merging 6c9b9e38e837893d36c18fdcb3b4b6ed75a8a2e1 into 3ccab6b1ff69584107ed5b488e542889b379852d - [view on LGTM.com](https://lgtm.com/projects/g/teamdigitale/io-app/rev/pr-b7b687b1337cad13abbb9aac32eb851e545fe1f3)

**new alerts:**

* 1 for Unused variable, import, function or class

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: ts/screens/authentication/IdpLoginScreen.tsx [hunk 1]
```diff
@@ -26,14 +27,24 @@ type OwnProps = {
 
 type Props = ReturnType<typeof mapStateToProps> & ReduxProps & OwnProps;
 
+/**
+ * web request state can be viewed in one of these states:
```
Comment (cloudify) [inline ts/screens/authentication/IdpLoginScreen.tsx]: check this comment, there are a few typos

File: ts/screens/authentication/IdpLoginScreen.tsx [hunk 2]
```diff
@@ -26,14 +27,24 @@ type OwnProps = {
...
+ * 'loading' is the initial state, cause the webview receive the end point url directly from props
+ * 'completed' means the request has been sucessfully
+ * 'error' means the request got an error (end point unrechable, timeout, http errors (404,500...))
+ */
+type RequestState = "loading" | "completed" | "error";
```
Comment (cloudify) [inline ts/screens/authentication/IdpLoginScreen.tsx]: we usually use all caps for _kinds_

File: ts/screens/authentication/IdpLoginScreen.tsx [hunk 3]
```diff
@@ -105,23 +140,68 @@ class IdpLoginScreen extends React.Component<Props, State> {
...
-          javaScriptEnabled={true}
-          onNavigationStateChange={handleNavigationStateChange}
-        />
-        {isWebViewLoading && (
-          <View style={styles.refreshIndicatorContainer}>
-            <RefreshIndicator />
-          </View>
+        {!hasError && (
+          <WebView
+            source={{ uri: "http://somedomain.doesnt.exists" }}
```
Comment (cloudify) [inline ts/screens/authentication/IdpLoginScreen.tsx]:
```ts
// TODO: remove this before merge
```

**Comment by digitalcitizenship** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  Please include a description of your PR changes.
  </td>
    </tr>
  </tbody>
</table>



## Affected stories

  * 🌟 [#165180769](https://www.pivotaltracker.com/story/show/165180769): Come CIT, se durante il login SPID fallisce il caricamento della webview, voglio che la app mi mostri un messaggio comprensibile.

<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

File: ts/screens/authentication/IdpLoginScreen.tsx [hunk 4]
```diff
@@ -94,9 +122,18 @@ class IdpLoginScreen extends React.Component<Props, State> {
     }
     const loginUri = LOGIN_BASE_URL + loggedOutWithIdpAuth.idp.entityID;
 
+    const handleOnError = (): void => {
```
Comment (cloudify) [inline ts/screens/authentication/IdpLoginScreen.tsx]: you can avoid the curly braces here

File: ts/screens/authentication/IdpLoginScreen.tsx [hunk 5]
```diff
@@ -94,9 +122,18 @@ class IdpLoginScreen extends React.Component<Props, State> {
...
+      this.setState({
+        requestState: "ERROR"
+      });
+    };
+
+    const goBack = () => this.props.navigation.goBack();
```
Comment (cloudify) [inline ts/screens/authentication/IdpLoginScreen.tsx]:
```suggestion
    const goBack = this.props.navigation.goBack;
```

File: ts/screens/authentication/IdpLoginScreen.tsx [hunk 6]
```diff
@@ -94,9 +122,18 @@ class IdpLoginScreen extends React.Component<Props, State> {
...
+    const refresh = () => this.setState({ requestState: "LOADING" });
```
Comment (cloudify) [inline ts/screens/authentication/IdpLoginScreen.tsx]:
```suggestion
    const setRequestStateToLoading = () => this.setState({ requestState: "LOADING" });
```

File: ts/screens/authentication/IdpLoginScreen.tsx [hunk 7]
```diff
@@ -26,14 +26,24 @@ type OwnProps = {
 
 type Props = ReturnType<typeof mapStateToProps> & ReduxProps & OwnProps;
 
+/**
+ * web request can be viewed in one of these states:
+ * 'loading' is the initial state, the webview receives the url directly from props so it starts to load
+ * 'completed' means the request has been successfully executed
+ * 'error' means the request got an error (endpoint unreachable, timeout, http errors (404,500...))
+ */
+type RequestState = "LOADING" | "COMPLETED" | "ERROR";
```
Comment (cloudify) [inline ts/screens/authentication/IdpLoginScreen.tsx]: you can probably use a `Pot` here as it covers all these states: https://github.com/teamdigitale/io-ts-commons/blob/master/src/pot.ts

File: locales/en/index.yml [hunk 8]
```diff
@@ -174,6 +174,11 @@ authentication:
   errors:
     notoken: The session is not ready or is invalid
     logout: Error on logout
+    network: 
+      title: Access problem
+      body: The identity provider chosen is not available at this time, you can try again later.
```
Comment (cloudify) [inline locales/en/index.yml]:
```suggestion
      body: The selected identity provider is not available at this time, you can try again later.
```

File: ts/screens/authentication/IdpLoginScreen.tsx [hunk 9]
```diff
@@ -94,9 +116,20 @@ class IdpLoginScreen extends React.Component<Props, State> {
...
+
+    const goBack = this.props.navigation.goBack;
+    const setRequestStateToLoading = () =>
+      this.setState({ requestState: pot.noneLoading });
+
     const handleNavigationStateChange = (event: NavState): void => {
       this.setState({
-        isWebViewLoading: event.loading ? event.loading : false
+        requestState: event.loading
+          ? pot.noneLoading
```
Comment (fpersico) [inline ts/screens/authentication/IdpLoginScreen.tsx]: Are you sure about this? Why you are sure it is noneLoading and not someLoading?

File: ts/screens/authentication/IdpLoginScreen.tsx [hunk 10]
```diff
@@ -94,9 +116,20 @@ class IdpLoginScreen extends React.Component<Props, State> {
     }
     const loginUri = LOGIN_BASE_URL + loggedOutWithIdpAuth.idp.entityID;
 
+    const handleOnError = (): void =>
```
Comment (fpersico) [inline ts/screens/authentication/IdpLoginScreen.tsx]: Why this function is declared inside render method?

Comment (fpersico) [reply ts/screens/authentication/IdpLoginScreen.tsx]: Sam for all the others.

File: ts/screens/authentication/IdpLoginScreen.tsx (hunk 9 above)
Comment (Undermaken) [reply ts/screens/authentication/IdpLoginScreen.tsx]: @fpersico yes, someLoading has more sense when webview changes state

File: ts/screens/authentication/IdpLoginScreen.tsx [hunk 11]
```diff
@@ -102,6 +102,71 @@ class IdpLoginScreen extends React.Component<Props, State> {
...
+        : pot.some("loading complete")
+    });
+
+    onNavigationStateChange(
+      () => this.props.dispatch(loginFailure()),
+      token => this.props.dispatch(loginSuccess(token))
+    )(event);
+  };
+
+  public renderMask = () => {
```
Comment (cloudify) [inline ts/screens/authentication/IdpLoginScreen.tsx]: shouldn't this be `private` too?

**Comment by cloudify** [comment]
@Undermaken 
pls add a screenshot to let @matteodesanti check that everything is correct before we merge

**Comment by Undermaken** [comment]
@matteodesanti 
## before errors handling

before


## after error handling
after

**Pull request merged**